modules/ssl/ssl_engine_kernel.c

```diff
@@ -2465,6 +2473,40 @@ int ssl_callback_ServerNameIndication(SSL *ssl, int *al, modssl_ctx_t *mctx)
 }
 
 #if OPENSSL_VERSION_NUMBER >= 0x10101000L && !defined(LIBRESSL_VERSION_NUMBER)
+/*
+ * Copy data from clienthello for env vars use later
+ */
+static void copy_clienthello_vars(conn_rec *c, SSL *ssl)
+{
+    SSLConnRec *sslcon;
+    modssl_clienthello_vars *clienthello_vars;
+    const unsigned char *data;
+    int *ids;
+
+    sslcon = myConnConfig(c);
+
+    sslcon->clienthello_vars = apr_palloc(c->pool, sizeof(*clienthello_vars));
+    clienthello_vars = sslcon->clienthello_vars;
+
+    clienthello_vars->version = SSL_client_hello_get0_legacy_version(ssl);
+    clienthello_vars->ciphers_len = SSL_client_hello_get0_ciphers(ssl, &data);
+    clienthello_vars->ciphers_data = apr_pmemdup(c->pool, data, clienthello_vars->ciphers_len);
+    if (SSL_client_hello_get1_extensions_present(ssl, &ids, &clienthello_vars->extids_len) == 1) {
+        clienthello_vars->extids_data = apr_pmemdup(c->pool, ids, clienthello_vars->extids_len * sizeof(int));
```

Comment:
Does this work if ->extids_len == 0 can be returned succesfully?

Also, to avoid a possible dangling ->extids_data I would either:
```
else {
    clienthello_vars->extids_data = NULL;
    clienthello_vars->extids_len = 0;
}
```
or use apr_pcalloc() above, the additional/implicit memset() would be in the noise in this path anyway (I think), possibly more future extensions proof too.

Reply:
good catch. We have a big issue if extids_len is never set.

Reply:
There were a couple serious bugs and other issues here. Much of this code assumed that the openssl functions would set the associated lengths correctly even if the functions returned failure which is not right. 

To fix these issues, I've rewritten this whole code block doing the following:
 - using calloc for struct (strictly shouldn't be necessary but agreed this is not expensive)
 - correctly check return codes from various SSL_client_hello_* functions, explicitly set len vars to 0 on failure
 - conditions to check len > 0 before memdup, explicitly set pointers to null if len == 0
 
I also have tested with diversity of clients including some lacking some collected extensions that prior to fixes would cause crash and other undefined behavior. These cases were all handled correctly after fixes.